code/aind_ccf_reg/preprocess.py

```diff
@@ -0,0 +1,296 @@
+        start_time = datetime.now()
+        low_thresh = threshold_li(arr_img)
+        end_time   = datetime.now()
+
+        logger.info(
+            f"Find optimal threshold using Li thresholding, execution time: {end_time - start_time} s -- low_thresh={low_thresh}"
+        )    
+        return low_thresh
+
+    def _cleanup_mask(self, arr_mask: np.ndarray) -> np.ndarray:
+        """
+        Morphological operations will be applied to clean up the mask by closing holes and
+        eroding away small or weakly-connected areas. The following steps are applied:
+            - Closing holes
+            - Dilation with radius 1 voxel
+            - Morphological closing
+            - Retain largest component
+        """
+        # 3x3 structuring element with connectivity 2
+        struct = ni.generate_binary_structure(3, 2)
+
+        mask = ni.binary_fill_holes(arr_mask).astype(int)
+        mask = ni.binary_dilation(mask, structure=struct).astype(int)
+        mask = ni.binary_closing(mask).astype(int)
+        mask = self._getLargestCC(mask)   
+
+        mask = ni.binary_dilation(mask, structure=struct, iterations=6).astype(int)
+        arr_mask = ni.binary_fill_holes(mask).astype(int)
+
+        return arr_mask
+
+
+    def run(self) -> np.ndarray:
+        """ compute the mask """
+        arr_img = self.ants_img.numpy()
+
+        # get optimal threshold using Li thresholding
+        # https://scikit-image.org/docs/stable/auto_examples/developers/plot_threshold_li.html
+        low_thresh = self._get_threshold_li(arr_img)
+
+        # thresholding
+        arr_mask = arr_img > low_thresh
+
+        # clean up
+        arr_mask = self._cleanup_mask(arr_mask)
+
+        # convert numpy array to ants image
+        ants_img_mask = ants.from_numpy(
+                                arr_mask.astype("float32"), 
+                                spacing=self.ants_img.spacing,
+                                origin=self.ants_img.origin, 
+                                direction=self.ants_img.direction)
+
+        return ants_img_mask
+
+
+class Preprocess():
+    """
+    Class to Preprocess lightsheet data
+    1. resample to isotropic to have same resolution of the SPIM template
+    2. N4 bias correction
+    3. intensity normalization 
+    """
+    def __init__(self, args, input_data, reference_data):
+        self.args = args
+        self.input_data = input_data
+        self.reference_data = reference_data
+            
+    def resample(self, ants_img, ants_template):    
+        """ Resample OMEZarr image to the resolution of template """
+        logger.info(f"Resample OMEZarr image to the resolution of template")
+        ants_img = ants.resample_image(
+            ants_img, ants_template.spacing, False, 1
+        )
+
+        logger.info(f"Resampled OMEZarr dataset: {ants_img}")
+                
+        # #------------- TODO: do we need? -------------#
```

Comment:
If we're changing the .tiff file CCF image (which I think we're doing), then we need to remove this.